### DISCUSSION TITLE
Is it possible to customize the number of segments to be indexed in milvus?

### DISCUSSION THREAD

**Discussion by wookiist** [milvus-io/milvus#31374] [Q&A and General discussion] [answered]
Is it possible to customize the number of segments to be indexed in milvus?
If so, I'm also wondering if it's possible to control the concurrency of the segments.

**Comment by wookiist** [comment]
Also, I was wondering how much performance degradation occurs when indexing multiple collections simultaneously. Is there a metric that measures something?

**Comment by xiaofan-luan** [comment]
> Is it possible to customize the number of segments to be indexed in milvus? If so, I'm also wondering if it's possible to control the concurrency of the segments.

1. each indexnode take only 1 task each time, but it fully utilize the cpu as much as possible. So it's better to isolate index node in a separate Pod
2. multiple collection index building task will be queued. For 1M data point, index building takes 10s for IVF and ScaNN and 30-60s for graph index. Adding more index is definitely the way to improve that

**Reply by wookiist** [reply]
@xiaofan-luan 
Thanks, so assuming there is only one collection, am I correct in understanding that no matter how many indexnodes there are, only one indexnode is fully utilized? 
And if the number of indexnodes and collections is 1:1, indexing of each collection will be done by holding each indexnode one by one?

**Reply by yhmo** [reply, marked as answer]
Milvus manages data by segments. Each segment has an independent index.
Once a new segment is generated, the datacoord will assign a task to an indexnode. A task is to build the index for one segment.
By default, one indexnode can only execute one task at a time. There is a configuration in milvus.yaml to change the default value:
```
indexNode:
  scheduler:
    buildParallel: 1
```

Let's say you insert lots of data in a short time, into a collection, 4 segments generated. There are 3 index nodes in the cluster.
The datacoord will assign task_1 to node_1, task_2 to node_2, task_3 to node_3. The 3 tasks are executed in the same time.
If the task_1 is completed first, the datacoord will assign task_4 to node_1.

**Reply by xiaofan-luan** [reply]
doesn't make sense to do so.
1 segment could use up all cpus on indexnnode
